gdb/riscv-tdep.c

```diff
@@ -158,35 +158,44 @@ static const struct register_alias riscv_register_aliases[] =
 #undef DECLARE_CSR
 };
 
-static const gdb_byte *
-riscv_breakpoint_from_pc (struct gdbarch *gdbarch,
-			  CORE_ADDR      *bp_addr,
-			  int            *bp_size)
-{
-  /* TODO: Support C.EBREAK for compressed (16-bit) insns.  */
-  /* TODO: Support NOPs for >=6 byte insns.  */
-  static const gdb_byte sbreak_insn[] = { 0x73, 0x00, 0x10, 0x00, };
-
-  *bp_size = 4;
-
-  return sbreak_insn;
-}
-
 static int
 riscv_breakpoint_kind_from_pc (struct gdbarch *gdbarch, CORE_ADDR *pcptr)
 {
-  /* TODO: Support C.EBREAK for compressed (16-bit) insns.  */
-  /* TODO: Support NOPs for >=6 byte insns.  */
-  return 4;
+  if (gdbarch_tdep (gdbarch)->supports_compressed_isa == UNKNOWN) {
+    /* TODO: Because we try to read misa, it is not possible to set a
+     * breakpoint before connecting to a live target. A suggested workaround is
+     * to look at the ELF file in this case. */
+    struct frame_info *frame = get_current_frame ();
+    uint32_t misa = get_frame_register_unsigned (frame, RISCV_CSR_MISA_REGNUM);
+    if (misa & (1<<2))
+      gdbarch_tdep (gdbarch)->supports_compressed_isa = YES;
+    else
+      gdbarch_tdep (gdbarch)->supports_compressed_isa = NO;
+  }
+
+  if (gdbarch_tdep (gdbarch)->supports_compressed_isa == YES) {
+    return 2;
+  } else {
+    return 4;
+  }
 }
 
 static const gdb_byte *
 riscv_sw_breakpoint_from_kind (struct gdbarch *gdbarch, int kind, int *size)
 {
   /* TODO: Support C.EBREAK for compressed (16-bit) insns.  */
```

Comment:
I omitted to delete the cbreak comment.
I intended to delete the long instruction one, because no special handling is needed for those. Just overwite the start of the instruction with whatever size breakpoint and everything will just work.